allow WFS GetFeature upon multiple layers, fixes #16509

cmangeat opened the pull request:
see https://issues.qgis.org/issues/16509.

Don't wether it is possible to produce a tag for one byte.

cmangeat committed:
fix for https://issues.qgis.org/issues/16509

luipir commented:
@cmangeat is it possible to change PR title to be descriptive. The title will enter in the git history and would be better to have a understandable test instead of only a reference to an issue.
e.g set tile "do something: fixes #16509". In this way, when merged, the PR will close automatically the issue.

luipir commented:
tnx @cmangeat

cmangeat committed:
wfs, get features upon many layers, add test

and fix iterator trouble (avoid bad values in response summary, wfs:FeatureCollection, TYPENAME)

cmangeat committed:
try to please indentation test

m-kuhn commented:
Great work that you have put into this unit test. I'd love to see more PRs like this!

> Don't wether it is possible to produce a tag for one byte.

What are you referring to with "tag"? A patch release?

@elpaso has a lot of experience with server tests. Would

cmangeat commented:
Can I do something in order to help this request to be merged ?

cmangeat committed:
use unitest assert and os.path.join as adviced

m-kuhn commented:
Thanks, @elpaso could you also have a look at this? Thanks!

elpaso commented on src/server/qgswfsserver.cpp:
```diff
@@ -445,7 +445,7 @@ int QgsWFSServer::getFeature( QgsRequestHandler& request, const QString& format
     for ( int i = 0; i < queryNodes.size(); i++ )
     {
-      queryElem = queryNodes.at( 0 ).toElement();
+      queryElem = queryNodes.at( i ).toElement();
```
Hi, I know this is unrelated with your modifications, but if I'm not wrong, this first loop at line 446 is dead code: all what it does it setting `mTypeNames` that it's not used and it's being redefined around line 878.
Would you please have a look and remove the loop if it's dead?

elpaso replied:
Travis is still failing for an indentation error in the Python code, would you please fix it so that we can finally merge?
See the error at:
https://travis-ci.org/qgis/QGIS/jobs/260609197#L1202

Thanks for your patience.

cmangeat replied:
Hi, I don't know exactly how it works, but I remember that mTypeNames preparation around line 448 were important and helps setting "summary" in xml/http response. Removing it cause missing "TYPENAME", e.g.

"""<wfs:FeatureCollection xmlns:wf...=1.0.0&amp;REQUEST=DescribeFeatureType&amp;TYPENAME=&amp;OUTPUTFORMAT=XMLSCHEMA"><gml:boundedBy>"""

instead of  

"""<wfs:FeatureCollection xmlns:wf...=1.0.0&amp;REQUEST=DescribeFeatureType&amp;TYPENAME=secondlayer,testlayer&amp;OUTPUTFORMAT=XMLSCHEMA"><gml:boundedBy>"""

elpaso reviewed (approved):
Thanks for the patch and for the test: I've restarted Travis to see if now passes because the failure was unrelated.
I've left a short comment in the code, would you please have a look?

cmangeat committed:
fix indentaton test  ? (scripts/prepare-commit.sh)

elpaso commented:
Thanks! Did you check if this needs forward porting to master?

cmangeat commented:
There is no need for forward porting, seems there was a big refactoring / improvement work since.
